src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2022 Wrenna Robson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wrenna Robson
+-/
+
+import analysis.normed_space.basic
+
+/-!
+# Hamming spaces
+
+The Hamming metric counts the number of places two members of a (finite) Pi type
+differ. The Hamming norm is the same as the Hamming metric over additive groups, and
+counts the number of places a member of a (finite) Pi type differs from zero.
+
+This is a useful notion in various applications, but in particular it is relevant
+in coding theory, in which it is fundamental for defining the minimum distance of a
+code.
+
+In this file we define `hamm β`, the type synonym of a Pi type with the Hamming
+distance `hamm_dist` and weight `hamm_wt` attached, and the various instances that arise
+from the properties of these definitions.
+
+-/
+
+/-
+We define `hamm_dist` and `hamm_wt` over Pi types, and will later attach them to our type
+synonym.
+-/
+section hamm_dist_wt
+
+open fintype function
+
+variables {α ι : Type*} {β : ι → Type*} [fintype ι] [Π i, decidable_eq (β i)]
+variables {γ : ι → Type*} [Π i, decidable_eq (γ i)]
+
+/-- The Hamming distance function to the naturals. -/
+def hamm_dist (x y : Π i, β i) := card {i // x i ≠ y i}
+
+lemma hamm_dist_nonneg {x y : Π i, β i} : 0 ≤ hamm_dist x y := zero_le _
+
+lemma hamm_dist_le_card_fintype {x y : Π i, β i} : hamm_dist x y ≤ card ι := card_subtype_le _
+
+lemma hamm_dist_comp_le_hamm_dist (f : Π i, γ i → β i) {x y : Π i, γ i} :
+  hamm_dist (λ i, f i (x i)) (λ i, f i (y i)) ≤ hamm_dist x y :=
+card_subtype_mono _ _ (λ _ H1 H2, by { simp_rw H2 at H1, exact H1 rfl })
+
+lemma hamm_dist_comp (f : Π i, γ i → β i) {x y : Π i, γ i} (hf : Π i, injective (f i))
+  : hamm_dist (λ i, f i (x i)) (λ i, f i (y i)) = hamm_dist x y :=
+le_antisymm (hamm_dist_comp_le_hamm_dist  _) (card_subtype_mono _ _ (λ i H1 H2, H1 (hf i H2)))
+
+lemma hamm_dist_smul_le_hamm_dist [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i} :
+  hamm_dist (k • x) (k • y) ≤ hamm_dist x y :=
+hamm_dist_comp_le_hamm_dist (λ i (c : β i), k • c)
+
+lemma hamm_dist_smul [Π i, has_scalar α (β i)] {k : α} {x y : Π i, β i}
+  (hk : Π i, is_smul_regular (β i) k) : hamm_dist (k • x) (k • y) = hamm_dist x y :=
+hamm_dist_comp (λ i (c : β i), k • c) hk
+
+@[simp] lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
+by {  rw [function.funext_iff, hamm_dist, card_eq_zero_iff],
+      exact ⟨ λ h i, imp_of_not_imp_not _ _ (λ H, h.elim' ⟨i, H⟩) h,
+              λ h, subtype.is_empty_of_false (λ i H, H (h _))⟩ }
```

Comment:
We don't usually indent like this in mathlib. Either way, we're missing a lemma here:
```lean
lemma subtype.is_empty_iff {α : Sort*} (p : α → Prop) : is_empty (subtype p) ↔ ∀ x, ¬p x :=
⟨λ h i hi, h.elim' $ subtype.mk i hi, subtype.is_empty_of_false⟩

@[simp] lemma hamm_dist_eq_zero {x y : Π i, β i} : hamm_dist x y = 0 ↔ x = y :=
by simp_rw [function.funext_iff, hamm_dist, card_eq_zero_iff, subtype.is_empty_iff, not_ne_iff]
```

I'll add it in a separate PR since it lives so far down the import heirarchy.

Reply:
And then we just put this PR on top?

Reply:
We do have `is_empty_iff`, not in the subtype namespace. You can do it with that.